consensus/src/processes/pruning_proof/mod.rs

```diff
@@ -614,45 +791,267 @@ impl PruningProofManager {
+        pp_header
+            .parents_by_level
+            .iter()
+            .enumerate()
+            .skip(1)
+            .find_map(|(level, parents)| {
+                if BlockHashSet::from_iter(parents.iter().copied()) == direct_parents {
+                    None
+                } else {
+                    Some((level - 1) as BlockLevel)
+                }
+            })
+            .unwrap_or(self.max_block_level)
+    }
+
+    fn estimated_blue_depth_at_level_0(&self, level: BlockLevel, level_depth: u64, current_dag_level: BlockLevel) -> u64 {
+        level_depth.checked_shl(level.saturating_sub(current_dag_level) as u32).unwrap_or(level_depth)
+    }
+
+    /// selected parent at level = the parent of the header at the level
+    /// with the highest blue_work (using score as work in this case)
+    fn find_selected_parent_header_at_level(
+        &self,
+        header: &Header,
+        level: BlockLevel,
+    ) -> PruningProofManagerInternalResult<Arc<Header>> {
+        // Parents manager parents_at_level may return parents that aren't in relations_service, so it's important
+        // to filter to include only parents that are in relations_service.
+        let parents = self
+            .parents_manager
+            .parents_at_level(header, level)
+            .iter()
+            .copied()
+            .filter(|parent| self.level_relations_services[level as usize].has(*parent).unwrap())
+            .collect_vec()
+            .push_if_empty(ORIGIN);
+
+        let mut sp = SortableBlock {
+            hash: parents[0],
+            blue_work: if parents[0] == ORIGIN { 0.into() } else { self.headers_store.get_header(parents[0]).unwrap().blue_work },
+        };
+        for parent in parents.iter().copied().skip(1) {
+            let sblock = SortableBlock {
+                hash: parent,
+                blue_work: self
+                    .headers_store
+                    .get_header(parent)
+                    .unwrap_option()
+                    .ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof(format!(
+                        "find_selected_parent_header_at_level (level {level}) couldn't find the header for block {parent}"
+                    )))?
+                    .blue_work,
+            };
+            if sblock > sp {
+                sp = sblock;
+            }
+        }
+        // TODO: For higher levels the chance of having more than two parents is very small, so it might make sense to fetch the whole header for the SortableBlock instead of blue_score (which will probably come from a compact header).
+        self.headers_store.get_header(sp.hash).unwrap_option().ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof(
+            format!("find_selected_parent_header_at_level (level {level}) couldn't find the header for block {}", sp.hash,),
+        ))
+    }
```

Comment:
I believe the following captures the same logic precisely 

```suggestion
    fn find_selected_parent_header_at_level(
        &self,
        header: &Header,
        level: BlockLevel,
    ) -> PruningProofManagerInternalResult<Arc<Header>> {
        // Parents manager parents_at_level may return parents that aren't in relations_service, so it's important
        // to filter to include only parents that are in relations_service.
        let sp = self
            .parents_manager
            .parents_at_level(header, level)
            .iter()
            .copied()
            .filter(|p| self.level_relations_services[level as usize].has(*p).unwrap())
            .filter_map(|p| self.headers_store.get_header(p).unwrap_option().map(|h| SortableBlock::new(p, h.blue_work)))
            .max()
            .ok_or(PruningProofManagerInternalError::NotEnoughHeadersToBuildProof("no parents with header".to_string()))?;

        Ok(self.headers_store.get_header(sp.hash).expect("unwrapped above"))
    }
```

Reply:
it's a bit more forgiving in the sense that it allows a parent with no header if others have a header, but I see no reason for this code to no allow that (and it's an unexpected scenario anyway: only origin has relations w/o header and origin would be a single parent by def)